[WIP] Data-structures tutorial (closes #57)

mikeal opened the pull request:
This is not finished but feedback is welcome.

vmx commented:
I've read through the markdown files and I don't have more to say than: great work!

terichadbourne commented:
Addresses issue #57

terichadbourne commented:
I'd love to discuss this in more detail with you when you're back, @mikeal, but here are some first impressions and questions after a quick read-through of the current draft of this workshop, shared in the interest of making this content as beginner-friendly as possible: 

1. When introducing data structures in general (lesson 1), it would be great to mention some common examples that developers will be more familiar with. Of those listed in the Wikipedia article, I think arrays, objects, and graphs would be the most recognizable to a web developer. I'm suggesting mentioning them in passing to make the reader realize they already have some fundamental related knowledge, not going into any detail.

2. `Decentralized data-structures are collections of data-structures that are linked to one another.` Is it really true that linked data structures are decentralized, or is it more accurate to say that linked data structures are particularly well-suited for decentralized networks? (Lesson 1)  Is the reason that they're well suited (for cases like git) the linked-ness or the consistent hashing so you're guaranteed everyone's using the same data? Or are both equally important?

3. `This allows one data-structure to be distributed over many different computers and protocols.` Is it really true that the _data structure_ is distributed, or is it that the _data_ is distributed over / copied onto many computers, and the _data structure_ is the key to ensuring we all know how to interpret that distributed data? (lesson 1) Note that I'd feel differently if you'd written "structured data" instead of "data structures" here. 

4.  I don't know of any reason to hyphenate "data structures," and it's not hyphenated in the Wikipedia article. Are you trying to convey a new meaning by hyphenating it that I'm missing?

5. When learning new concepts, I find that contrasting them to things I'm already familiar with is really helpful. Two opportunities for us to use this teaching method here would be to specifically name and describe **centralized networks** (versus decentralized networks) and **location-based addressing** (versus content-based addressing). Calling out these contrasts would be an opportunity to highlight the "why" behind decentralized (security, etc).

6. I'm not sure it's fair to assume that people know what a hash is. (lesson 1)  If you're not familiar with cryptography and you know Ruby, you might think it's the equivalent of a JavaScript object, for example. I'd love to see more here about how you're turning your photo of a kitten into a string of characters (hash) that represents that picture of a kitten, emphasizing that every time you do it to the same pic, you get the same hash back. 

7. It's not yet clear that when you change a single pixel of that kitten photo, the hash changes. This could be addressed somewhere in Lesson 2. Perhaps Bob changed a word/pixel in the doc he received from Alice, for example, so now it would have a new hash. You have to understand that a single change to the data changes its hash before you can understand the statement in lesson 3 that `As this tree above is built the final content address (hash) of the root node is unique to a tree than contains every node all the way down this tree.`

8. You may just not have gotten to it yet, but I hope to see more about what DAG means, particularly since all the commands we see later include `dag` in them. My impression from reading elsewhere is that IPFS is a combination of merkle-ness and dag-ness.

9. Should definitions of and differences between IPFS and IPLD be discussed in this lesson? IPFS is mentioned once with no description of what it is. IPLD is never mentioned that I've spotted. These definitions could (a) be mixed into the content of these existing lessons, (b) be another lesson in this series, or (c) be another series that comes after this one and before people start working with the IPFS commands in the Basics series. 

10. Should we say something about what formats of data can be stored in IPFS / decentralized data structures? Does this hashing system work equally well for pictures and text documents and JSON objects? Does it somehow store the file extension so you can figure out how to turn it back into a kitten picture on the other end? This might go in the IPFS/IPLD section as opposed to specifically the more general section, if we do make a distinction there. 

Happy to answer any questions about my questions when you're back.

terichadbourne commented:
Just spotted your convo with Paul about decentralized data structures and there may be some language there that we could pull in as well: https://twitter.com/pfrazee/status/1054431271986913280

terichadbourne commented:
Note that due to [a change I'm making elsewhere](https://github.com/ipfs-shipyard/proto.school/pull/63/commits/35a329510f134467b49b8723d99aab0aa53049a1), we'll need to update the last lesson in your series (once we know which that is) as follows to ensure it links to the workshop list instead of a non-existent subsequent lesson: 
- Add `last="true"` to the values being passed to the Lesson component.  (The button will now populate automatically.)
- Use the success message "Great job! You've completed this series of lessons!" or something else that makes it clear this was the last lesson in the series.

terichadbourne commented:
@mikeal How would you like to approach finishing up this lesson? Do you want to hop on a call and chat through the first impressions I've shared here?

terichadbourne commented:
With a revision I'm making today in PR #83, you'll no longer have to add `last: true` as I'd previously indicated.

mikeal commented:
Pushed some new stuff. Responses to your earlier points:

1. Added a sentence.
2. Write a new section on verifiability and re-worded this sentence.
3. It is actually true that, quite literally, the structure of the data can be spread over multiple peers. Each peer may only have part of the data and the organizational structure of the data is only clear if you have all the data. Re-worded this sentence a bit tho.
4. Fixed.
5. We contrast URL (centralized addressing) with content addressing in the second lesson. Is this not sufficient?
6. Added a link to Wikipedia. The TODO code sample should help here too.
7. Added more language to this effect in lesson 3.
8. DAG's *are* merkle trees :) I'll make sure to address that in the lesson.
9. Perhaps that's what the lesson should eventually close with.
10. We could talk about this but it very quickly gets into fairly advanced topics. I'm not sure if we want to go down this particular rabbit hole.

terichadbourne commented:
@mikeal I've gone through and reworked the existing lesson structure and content to something that feels more beginner-friendly to me. Could you please take a look to make sure I haven't introduced any inaccuracies? 

We had discussed the possibility of adding a final lesson introducing IPFS and IPLD, but I think you should be the one to do the first stab at that if we want to add it. 

I feel like the linking stuff is clear now, and it's obvious what content addressing has to do with the distributed web, but it's not super obvious why we're talking about merkle trees and DAGs. If you have ideas for fixing that, let me know! 

There's not currently much of a conclusion here, but that's in part due to my finding the last section the least easy to explain the importance of in this context. 

I'm comfortable merging this tutorial now, but after the holidays I'd like to ask the non-developers in the company to read through this whole tutorial and tell me what's confusing. I did my best to keep the language accessible to non-programmers.

terichadbourne commented:
@mikeal This PR to finish off milestone 1.0 is awaiting your approval. I'd appreciate you checking to ensure I haven't introduced any inaccuracies while trying to simplify the content of the new tutorial. Thanks!

terichadbourne commented:
This also now closes #107

mikeal commented on src/lessons/Data-Structures/01.md:
```diff
@@ -0,0 +1,33 @@
+from a central authority, we need specialized data structures that allow us to verify
+and link between various pieces of content.
+
+Data structures that are using decentralized systems need to be <strong>verifiable</strong>. On a single
```
You should be able to use markdown syntax here instead of html <strong>. Some markdown parsers are not forgiving about *any* HTML, even safe html like this.

mikeal commented on src/lessons/Data-Structures/01.md:
```diff
@@ -0,0 +1,33 @@
+or on disc. But in a decentralized system you have less, or possibly zero, trust among peers.
+
+Also, large data structures need to be able to be spread out among peers and <strong>linked together</strong>
+in order to be decentralized. In the same way that any web page on the web can link to another web page
```
Can just say "any web page," "any web page on the web" is a bit redundant.

mikeal commented on src/lessons/Data-Structures/03.md:
```diff
@@ -0,0 +1,56 @@
+
+Hashing takes data of any size and returns you a single fixed-size "hash" for that data.
+A hash is a string of characters that looks like gobbledygook, but you can think of it
+as a link. It might look something like this:
```
"think of it as a link" feels like we skipped a step. "think of it as a name" seems to be in-line with where the current narrative is at. Somewhere towards the end we can note "if we use these names as links..."

mikeal commented on src/lessons/Data-Structures/03.md:
```diff
@@ -0,0 +1,56 @@
+of those two photos is identical.
+
+<strong>Cryptographic hashes are also unique</strong>. That means that if Jess uses Photoshop to remove a
+single whisker from that kitty, the udpated image will have a new hash, and it will
```
```suggestion
single whisker from that kitty, the updated image will have a new hash, and it will
```

mikeal commented on src/lessons/Data-Structures/03.md:
```diff
@@ -0,0 +1,56 @@
+access to that image.
+
+The decentralized web works differently. When we want a specific photo of an adorable pet, we ask for it
+by its content address (hash). Who do we ask? The whole network! If Kyle is online, we'll see that he has
```
At this point I completely had forgotten about Jess and Kyle and had to scroll up to remember.

mikeal commented on src/lessons/Data-Structures/04.md:
```diff
@@ -0,0 +1,51 @@
+
+A `CID` is a single identifier that contains a cryptographic hash **and**
+information about how to interpret that data called a "codec." (Codecs encode,
+decode, compress, and decompress data in certain formats. You use them every time
```
I'd leave compress and decompress out for now. Technically, there's not compression in the current codecs, and we're trying to do compression at another layer (transport and storage). This will be quite confusing for anyone trying to map these semantics on to what they already know about HTTP so I'd just leave compression out for now.

mikeal commented on src/lessons/Data-Structures/04.md:
```diff
@@ -0,0 +1,51 @@
+A `CID` is a single identifier that contains a cryptographic hash **and**
+information about how to interpret that data called a "codec." (Codecs encode,
+decode, compress, and decompress data in certain formats. You use them every time
+you stream a YouTube video of an adorable beagle.)
```
Video codecs, like those on YouTube, are a different beast, so I'd leave this out.

In fact, we already have be aware of video codecs when encoding files in IPFS, just at a different layer (chunking).

mikeal commented on src/lessons/Data-Structures/04.md:
````diff
@@ -0,0 +1,51 @@
++-------+------------------------------+
+```
+
+Many formats and protocols use content addressing. `CID` allows us to create a universal identifier for any of these sytems. Tools like git and protocols like Etherium and Bitcoin all use content addressing already, but they differ greatly in how to interpret the data and what cryptographic function they use for hashing.
````
```suggestion
Many formats and protocols use content addressing. `CID` allows us to create a universal identifier for any of these systems. Tools like git and protocols like Ethereum and Bitcoin all use content addressing already, but they differ in how to interpret the data and what cryptographic function they use for hashing.
```

mikeal commented on src/lessons/Data-Structures/04.md:
```diff
@@ -0,0 +1,51 @@
+giraffes. The possiblities are endless!) All the way down this tree we
+have cryptographic hashes that allow us to distribute and link the data.
+
+Why is important to link between different data structures? Every day on the centralized web,
```
```suggestion
Why is it important to link between different data structures? Every day on the centralized web,
```

mikeal commented on src/lessons/Data-Structures/04.md:
```diff
@@ -0,0 +1,51 @@
+
+Why is important to link between different data structures? Every day on the centralized web,
+we link from text to images, from logos to homepages, and from emails to PDFs.
+Links ties resources together, convey meaning, and make the web awesomely interactive!
```
```suggestion
Links tie resources together, convey meaning, and make the web awesomely interactive!
```

mikeal commented on src/lessons/Data-Structures/05.md:
```diff
@@ -0,0 +1,60 @@
+
+Getting lost? Let's imagine this as a set of file folders. If we run folder E through our hashing algorithm
+while it contains subfolders F and G, the content-derived hash we get back is unique to its current content.
+If we remove folder G, it's like removing Jess removing that whisker from her kitten photo. The folder formerly known as E doesn't
```
```suggestion
If we remove folder G, it's like Jess removing that whisker from her kitten photo. The folder formerly known as E doesn't
```

mikeal commented:
This is great!

I'd say that the changes you made have made this tutorial 8-13x better than the original :)

I've added a bunch of small comments in-line as well as a few suggestions.
